zio-http/shared/src/main/scala/zio/http/endpoint/Endpoint.scala

```diff
@@ -30,6 +30,18 @@ import zio.http._
 import zio.http.codec._
 import zio.http.endpoint.Endpoint.{OutErrors, defaultMediaTypes}
 
+sealed trait AuthType {
+  type ClientRequirement
+}
+
+object AuthType {
+  type None = None.type
+
+  case object None   extends AuthType { type ClientRequirement = Unit                        }
+  case object Basic  extends AuthType { type ClientRequirement = Header.Authorization.Basic  }
+  case object Bearer extends AuthType { type ClientRequirement = Header.Authorization.Bearer }
```

Comment:
FYI, there could the some more use cases:
1. Custom header like `X-API-Key`. The header name can vary.
2. Combination of multiple auth types: I have some endpoints that support both Bearer and `X-API-Key`.
3. I don't have that kind of use case, but there might be needs for cookie based authentication, when the endpoint need to support plain web browsers.

Reply:
Thanks for the feedback. I want to get comments about the general design. But we can make auth type a FP DSL with fallback. And I want to add more options for sure. Maybe we can make it work that users can define custom auth types. We actually only need the ClientRequirement and a Codec fo the server. But this all depends on the `authorization` header.
A more generic solution would then also need to define the header type and the client requirement/codec would need to use this header type. This would still work. The question is, if it is worth the effort.

Reply:
Sure, take you time. I just wanted to inform my use cases that may affect the design, not to force you to add all the missing stuff right now :sweat_smile:

In general, I think replacing endpoint middleware with dedicated auth input is way more practical, because it was alway difficult to imagine a general endpoint middleware that takes into account server, client, and documentation with necessary type juggling that may affect all the type members.